scripts/_functions_sass.R

```diff
@@ -0,0 +1,438 @@
+    daterange = shiny_sass_daterange_picker(),
+    marker = sass_layer_file(path_sass_markers),
+  )
+}
+
+bslib_component_sass <- function() {
+  bslib:::component_dependency_sass_files()
+}
+
+shiny_sass <- function(bs_version) {
+  shiny:::shinyDependencySass(bs_version)
+}
+
+shiny_sass_selectize <- function(bs_version) {
+  shiny:::selectizeSass(bs_version)
+}
+
+shiny_sass_ionrangeslider <- function() {
+  sass::sass_layer(rules = shiny:::ionRangeSliderDependencySass())
+}
+
+shiny_sass_daterange_picker <- function() {
+  shiny:::datePickerSass()
+}
+
+#' Flattens the Sass bundle into a single Sass file, ignoring file attachments
+#' @return A character vector of lines of flattened Sass code
+theme_as_sass_lines <- function(full_theme) {
+  bs_sass <- as_sass(full_theme)
+  strsplit(as.character(bs_sass), "\n")[[1]]
+}
+
+#' Replaces package-relative paths in the theme Sass with paths relative to the
+#' `shiny/www/shared/sass/preset/{preset_name}` directory, where the preset Sass will
+#' be stored.
+theme_relative_paths <- function(theme_sass_lines) {
+  replace_package_path <- function(pkg) {
+    replacement <- "@import \"../../"
+
+    pkg_dir <- path_package(pkg)
+    if (path_file(pkg_dir) == "inst") {
+      # e.g /Users/garrick/work/rstudio/bslib/inst
+      replacement <- paste0('@import "../../', pkg, "/")
+    } else {
+      # e.g. /path/to/libPath/sass
+      pkg_dir <- path_dir(pkg_dir)
+    }
+
+    gsub(
+      paste0("@import \"", pkg_dir, "/"),
+      replacement,
+      theme_sass_lines,
+      fixed = TRUE
+    )
+  }
+
+  for (pkg in c("bslib", "shiny")) {
+    theme_sass_lines <- replace_package_path(pkg)
+  }
+
+  theme_sass_lines
+}
+
+#' Finds the Sass file for a given import statement, which may exclude a leading
+#' underscore and/or the file extension.
+#' @return The path to the Sass file, or `NULL` if the file could not be found.
+find_sass_file <- function(file) {
+  if (file_exists(file)) {
+    return(file)
+  }
+
+  file_scss <- path(file, ext = "scss")
+  file_sass <- path(file, ext = "sass")
+
+  if (file_exists(file_scss)) {
+    return(path(file_scss))
+  } else if (file_exists(file_sass)) {
+    return(path(file_sass))
+  } else if (!grepl("^_", path_file(file))) {
+    file_under <- path(path_dir(file), paste0("_", path_file(file)))
+    return(find_sass_file(file_under))
+  }
+
+  NULL
+}
+
+#' Searches the Sass lines recursively for `@import` statements and returns the paths to
+#' all imported files, including `@import` statements in imported files.
+#' @return A character vector of absolute paths to all imported Sass files.
+theme_files_used <- function(theme_sass_lines) {
+  imports <- grep("^\\s*@import \"", theme_sass_lines, value = TRUE)
+  if (!length(imports)) {
+    return()
+  }
+
+  imports <- trimws(imports)
+  imports <- gsub("^@import \"([^\"]+?)\";", 'identity("\\1")', imports)
+
+  files <- vapply(imports, \(x) eval(parse(text = x)), character(1))
```

Comment:
This is tricky to follow (and I think would miss single-quote imports?).

Would it be better to do something like `strsplit(imports, "@import", fixed = TRUE)`?

Reply:
This is a bit of a trick, but basically I'm turning `@import(value)` into `identity(value)` and then evaluating it as R code, returning whatever string is in the import statement, so both single and double quotes are handled.

Reply:
I simplified this and just used regular expressions rather than the sass -> R code shenanigans. 7cbd2d7a1c1d8164afab48f7db02ce0fbbcb33f1